tests/kafkatest/tests/replication_test.py

```diff
@@ -104,44 +152,84 @@ def run_with_failure(self, failure):
 
         assert success, msg
 
-    def clean_shutdown(self):
-        """Discover leader node for our topic and shut it down cleanly."""
-        self.kafka.signal_leader(self.topic, partition=0, sig=signal.SIGTERM)
+    def fetch_broker_node(self, topic, partition, node_type):
+
+        if node_type == LEADER:
+            return self.kafka.leader(topic=topic, partition=partition)
+        elif node_type == CONTROLLER:
+            return self.kafka.controller()
+        else:
+            raise RuntimeError("Unsupported node type.")
+
+    def start_producers(self):
+        for producer in self.producers:
+            producer.start()
+
+        for producer in self.producers:
+            if not wait_until(lambda: producer.num_acked > 5, timeout_sec=5):
+                raise RuntimeError("Producer failed to start in a reasonable amount of time: %s" % str(producer))
+
+    def stop_producers(self):
+        for producer in self.producers:
+            producer.stop()
+
+    def drive_failures(self, failure, node_type):
+
+        if failure in {SOFT_KILL, HARD_KILL, CLEAN_KILL}:
+            self.kill(failure, node_type)
+        elif failure in {SOFT_BOUNCE, HARD_BOUNCE, CLEAN_BOUNCE}:
+            self.bounce(failure, node_type, self.num_bounce)
+        else:
+            raise RuntimeError("Invalid failure type")
+
+    def bounce(self, failure, node_type, num_bounce):
+        for i in range(num_bounce):
+            node_to_signal = self.fetch_broker_node(self.topic_to_fail, 0, node_type)
+
+            if failure == CLEAN_BOUNCE or failure == HARD_BOUNCE:
+                self.kafka.restart_node(node_to_signal, wait_sec=5, clean_shutdown=(failure == CLEAN_BOUNCE))
+            elif failure == SOFT_BOUNCE:
+                self.kafka.signal_node(node_to_signal, "SIGSTOP")
+                time.sleep(20)
```

Comment:
This was a janky experiment to trigger reelection so that the previous leader could come back online after the new one was elected.

Replaced this with a more precise `wait_until(...)` call